### PR TITLE
Support OIDC Authentication

### PR DESCRIPTION
**Pull request by abellotti**
Adding support for OIDC Authentication in the Self-Service UI

Related PRs:
- https://github.com/ManageIQ/manageiq-appliance/pull/287
- https://github.com/ManageIQ/manageiq-api/pull/856
- https://github.com/ManageIQ/manageiq-documentation/pull/1460

Based upon & References:
- https://github.com/Plemi/manageiq-ui-service/commit/4908eb7edd512ff98d1514d9b223d24ed8df07dc
- https://github.com/zmartzone/mod_auth_openidc/issues/69
- https://github.com/zmartzone/mod_auth_openidc/wiki/Single-Page-Applications

### CODE REVIEW

**Comment by abellotti** [comment]
@miq-bot add_labels wip, enhancement

**Comment by abellotti** [comment]
With OpenID-Connect external authentication enabled, the Self-Service UI Login screen now appears as follows:

image

With the language selection as it exists in the default database authentication login screen.

**Comment by abellotti** [comment]
When the **Enable Single Sign-On** button is checked in the ManageIQ Authentication configuration page:

image


The Self-Service UI does not present the Language selection or the **Login to Corporate System** button:

image

It automatically redirects the user to the Identity Provider:

image


Similar to what is done with the Administrative UI.

**Comment by abellotti** [comment]
@miq-bot assign @himdel 

Ok @himdel, went through a couple of iterations and troubleshooting a couple of reliability issues, but I think I finally got it.  Seems to be behaving as expected. Please review when you get a chance.

/cc @jvlcek can you kick the Self-Service OIDC tires with this PR please ?

Thank you both.

File: client/app/core/authentication-api.factory.js [hunk 1]
```diff
@@ -10,13 +10,51 @@ function base64encode (str, encoding = 'utf-8') {
...
+        },
+      }).then(loginSuccess, loginFailure);
+
+      function loginSuccess (response) {
+        Session.setAuthToken(response.data.auth_token)
+        Session.setAuthMode(self.currentAuthMode)
+        resolve(response)
+      }
+
+      function loginFailure (response) {
```
Comment (himdel) [inline client/app/core/authentication-api.factory.js:47]: you don't need the extra copies of `loginSuccess` and `loginFailure` if they're the same

File: client/app/core/authentication-api.factory.js [hunk 2]
```diff
@@ -10,13 +10,51 @@ function base64encode (str, encoding = 'utf-8') {
...
+      self.currentAuthMode = authMode
+      return oidcLogin(access_token)
+    } else {
+      self.currentAuthMode = 'database'
+      return login(userLogin, password)
+    }
+  }
+
+  function oidcLogin (access_token) {
+    return new Promise((resolve, reject) => {
```
Comment (himdel) [inline client/app/core/authentication-api.factory.js:31]: this is not needed, `loginSuccess` can just `return response` to achieve the same result by just `return $http.get...`

File: client/app/core/session.service.js [hunk 3]
```diff
@@ -30,6 +35,20 @@ export function SessionFactory ($http, $localStorage, $cookies, RBAC, Polling, $
     $localStorage.token = model.token;
   }
 
+  function setAuthMode (authMode) {
+    if (validAuthModes.indexOf(authMode) !== -1) {
```
Comment (himdel) [inline client/app/core/session.service.js:39]: could be written as `!validAuthModes.includes(authMode)`

File: client/app/states/login/login.html [hunk 4]
```diff
@@ -1,4 +1,33 @@
-<div class="login-pf">
+<div class="login-pf" ng-if="vm.authenticationInfo && vm.authenticationInfo.oidc_enabled">
```
Comment (himdel) [inline client/app/states/login/login.html:1]: looks like the only differences are inside the `<form` element?

If so, we should probably move the ng-if there and not repeat the logos...

File: client/app/states/login/login.state.js [hunk 5]
```diff
@@ -101,10 +132,26 @@ function StateController ($window, $state, Text, RBAC, API_LOGIN, API_PASSWORD,
...
+        if (vm.oidcInitiatedLogin) {
+          vm.oidcInitiatedLogin = false
+          let idp_access_token = $cookies.get(oidc_access_token.name)
+          if (idp_access_token && idp_access_token != '') {
+            vm.access_token = idp_access_token
+            return AuthenticateUser()
+          }
+        } else {
+          if (vm.authenticationInfo.sso_enabled) {
+            return new Promise( () => { initiateOidcLogin() })
```
Comment (himdel) [inline client/app/states/login/login.state.js:151]: possibly a bug .. this returns a promise that never resolves or rejects

is this to run `initiateOidcLogin` asynchronously later? If so, maybe `$timeout(initiateOidcLogin)` would be clearer?

File: client/app/states/oidc/oidc-login.state.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  routerHelper.configureStates(getStates())
+}
+
+function getStates () {
+  return {
+    'oidc-login': {
+      parent: 'application',
+      url: '/oidc_login',
+      redirectTo: 'login',
+      template: '<ui-view></ui-view>'
```
Comment (himdel) [inline client/app/states/oidc/oidc-login.state.js:12]: redirectTo + template ... template is probably not needed?

**Comment by himdel** [comment]
I'm not sure I can test it, but except for a few minor issues above, the code looks good :+1:

File: client/app/core/authentication-api.factory.js (hunk 2 above)
Comment (abellotti) [reply client/app/core/authentication-api.factory.js:31]: 👍   much simpler. thank you !!

File: client/app/states/login/login.html (hunk 4 above)
Comment (abellotti) [reply client/app/states/login/login.html:1]: this also helped eliminate the period of time I had the screen white, before the ng-if expressions were valid. 👍

File: client/app/states/login/login.state.js (hunk 5 above)
Comment (abellotti) [reply client/app/states/login/login.state.js:151]: triggers the IDP login when sso is enabled, $timeout(initiateOidcLogin) worked great too 👍

**Comment by abellotti** [comment]
Thanks @himdel for the review (and all the Angular/JS help for this newbie :) ).  Once I push the changes @jvlcek will help out in the testing.

**Comment by jvlcek** [comment]
With @abellotti's help to patch an appliance with his configuration and testing we have confirmed proper functionality! 💯 👍 Good-to-go

**Comment by abellotti** [comment]
@miq-bot add_label jansa/yes?

**Comment by simaishi** [comment]
@abellotti @himdel Backporting this to jansa branch conflicts as jansa branch doesn't have https://github.com/ManageIQ/manageiq-ui-service/pull/1627. Not sure if I should backport that PR as well, or you'll need to create a separate PR for jansa. Please let me know. Thanks!

**Comment by himdel** [comment]
@simaishi I think you can backport #1627 as well, it makes sense to sync the error messages given the new authentication changes.

**Comment by simaishi** [comment]
Jansa backport details:

```
$ git log -1
commit ca1e6027d7e271aff0b76642cb50eeb69374d8f7
Author: Oleg Barenboim <chessbyte@gmail.com>
Date:   Fri Jun 19 10:28:12 2020 -0400

    Merge pull request #1651 from abellotti/sui-sso

    Support OIDC Authentication

    (cherry picked from commit 4236274edb236ed1b6f16d6c4f2f61a81f983669)
```

**Pull request merged**